### DISCUSSION TITLE
How to use Application.Current!.ActualThemeVariant in a converter ?

### DISCUSSION THREAD

**Discussion by flarive** [AvaloniaUI/Avalonia#11735] [Q&A] [answered]
Hello,

I need to get the actual theme variant inside a converter (i'm using Avalonia 11rc1)

```
object o;
Color shadowLightColor = default;
Color shadowDarkColor = default;

ThemeVariant theme = Application.Current!.ActualThemeVariant;

Application.Current!.TryGetResource("MaterialDesignShadowLightColor", theme, out o);
shadowLightColor = (Color)o;

Application.Current!.TryGetResource("MaterialDesignShadowDarkColor", theme, out o);
shadowDarkColor = (Color)o;
```

When i bind from xaml, the 2 shadow colors are not updated when i switch theme from light to dark.

```
<converters:ControlWidthToBoxShadowConverter x:Key="ControlWidthToBoxShadowConverter" />

<Border Name="PART_ButtonRootBorder" Margin="{TemplateBinding Margin}" Padding="0"
                Width="{TemplateBinding Width}" Height="{TemplateBinding Height}" Cursor="{TemplateBinding Cursor}"
                VerticalAlignment="Center" HorizontalAlignment="Center"
                Background="{TemplateBinding Background}"
                BorderBrush="{TemplateBinding BorderBrush}"
                BorderThickness="{TemplateBinding BorderThickness}"
                BoxShadow="{TemplateBinding Height, Converter={StaticResource ControlWidthToBoxShadowConverter}, ConverterParameter=0}"
                CornerRadius="{TemplateBinding CornerRadius}">
```

I have also tried to bind from {x:Static Avalonia:Application.Current.ActualThemeVariant} but it doesn't work.

Would would have an idea please ?

Many thanks

**Comment by robloo** [comment]
Pass it, or the entire Border control, as a parameter to the converter. I'm not sure why you are currently setting ConverterParameter=0 though -- seems unused.

**Comment by maxkatz6** [comment]
ConverterParameter or IMultiValueConverter

**Reply by flarive** [reply]
Ok, i have tried using a MultiValueConverter but i always get an unset value for my MultiBinding second parameter.

```
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                    ...
                    xmlns:naming="clr-namespace:Avalonia.Themes.Neumorphism.Resources.Naming">

<Border Name="PART_ButtonRootBorder" Margin="{TemplateBinding Margin}" Padding="0"
                Width="{TemplateBinding Width}" Height="{TemplateBinding Height}" Cursor="{TemplateBinding Cursor}"
                VerticalAlignment="Center" HorizontalAlignment="Center"
                Background="{TemplateBinding Background}"
                BorderBrush="{TemplateBinding BorderBrush}"
                BorderThickness="{TemplateBinding BorderThickness}"
                CornerRadius="{TemplateBinding CornerRadius}">

<Border.BoxShadow>
      <MultiBinding Converter="{StaticResource ControlWidthToBoxShadowConverter2}" ConverterParameter="0">
        <Binding Path="Height" RelativeSource="{RelativeSource TemplatedParent}"/>
        <Binding Path="{x:Static naming:PartNames.Test}"/>
      </MultiBinding>
</Border.BoxShadow>
```




```
namespace Avalonia.Themes.Neumorphism.Resources.Naming
{
    public static class PartNames
    {
        public static string Test => Application.Current.ActualThemeVariant.ToString();
    }
}

```

Don't understand why it doesn't work

**Reply by maxkatz6** [reply]
If you want to use x:Static inside of the binding, you need to use `Binding Source="{x:Static}"` instead, as it's not a path, but a whole object to bind as a source.
But I believe that's not what you want anyway, since PartNames won't trigger any changes (it can't have INotifyPropertyChanged). Instead you can bind to `Binding Source="{x:Static Application.Current.ActualThemeVariant}"` (untested, but should be fine).

**Reply by maxkatz6** [reply]
Also, this approach would be acceptable for the application, but since you are doing a third-party theme, I wouldn't recommend that.
Third-party themes should be usable with theme variants redefined in the middle of the tree. I.e. whole app can have a dark variant, but a single window (or a single control in the window) might have a light variant. Meaning, you shouldn't rely on Application.ActualThemeVariant.
Note, this wasn't possible before in 0.10 (at least reasonably easy), but is possible now with 11.0.

Please refer [to this PR](https://github.com/AvaloniaUI/Avalonia/pull/8166) as documentation for theme dictionaries (it will be copied to the actual documentation when it's live).
To solve your problem, I would change the way how this converter works. Instead of reading values from the CurrentTheme you could invert this dependency in a way like this:
```xaml
  <Border.BoxShadow>
        <MultiBinding Converter="{StaticResource ControlWidthToBoxShadowConverter2}" ConverterParameter="0">
          <Binding Path="Height" RelativeSource="{RelativeSource TemplatedParent}"/>
          <DynamicResource ResourceKey="ShadowMainColor" />
          <DynamicResource ResourceKey="ShadowRestColor" />
        </MultiBinding>
  </Border.BoxShadow>
```
And then define ShadowMainColor and ShadowRestColor with different values in different theme dictionaries. DynamicResource also works nicely inside of MultiBinding.

**Reply by maxkatz6** [reply]
In general, it would be better to define BoxShadow in the ThemeDictionary itself, as a normal resource. But I assume you need Height information from the control to calculate shadow properly.

Alternatively, you could define normal BoxShadow in the different theme resources without any Height information, but with proper colors, and then use DynamicResource + Converter to only apply height information to the specific control's shadow. Not sure if this solution is better than MultiBinding, but you can test it and see what works better for you.

**Reply by flarive** [reply]
> Also, this approach would be acceptable for the application, but since you are doing a third-party theme, I wouldn't recommend that. Third-party themes should be usable with theme variants redefined in the middle of the tree. I.e. whole app can have a dark variant, but a single window (or a single control in the window) might have a light variant. Meaning, you shouldn't rely on Application.ActualThemeVariant. Note, this wasn't possible before in 0.10 (at least reasonably easy), but is possible now with 11.0.
> 
> Please refer [to this PR](https://github.com/AvaloniaUI/Avalonia/pull/8166) as documentation for theme dictionaries (it will be copied to the actual documentation when it's live). To solve your problem, I would change the way how this converter works. Instead of reading values from the CurrentTheme you could invert this dependency in a way like this:
> 
> ```
>   <Border.BoxShadow>
>         <MultiBinding Converter="{StaticResource ControlWidthToBoxShadowConverter2}" ConverterParameter="0">
>           <Binding Path="Height" RelativeSource="{RelativeSource TemplatedParent}"/>
>           <DynamicResource ResourceKey="ShadowMainColor" />
>           <DynamicResource ResourceKey="ShadowRestColor" />
>         </MultiBinding>
>   </Border.BoxShadow>
> ```
> 
> And then define ShadowMainColor and ShadowRestColor with different values in different theme dictionaries. DynamicResource also works nicely inside of MultiBinding.

Such a good idea, many thanks !
I didn't think to do that like that, it's so simple, i should have think of it :)

**Reply by flarive** [reply]
Arg... New problem with MultiBinding :(

This works well :

```
<Border Name="PART_ButtonRootBorder" Margin="{TemplateBinding Margin}" Padding="0"
                Width="{TemplateBinding Width}" Height="{TemplateBinding Height}" Cursor="{TemplateBinding Cursor}"
                VerticalAlignment="Center" HorizontalAlignment="Center"
                Background="{TemplateBinding Background}"
                BorderBrush="{TemplateBinding BorderBrush}"
                BorderThickness="{TemplateBinding BorderThickness}"
		BoxShadow="{TemplateBinding Height, Converter={StaticResource ControlWidthToBoxShadowConverter}, ConverterParameter=0}"
                CornerRadius="{TemplateBinding CornerRadius}">

...

<!--Pointer enter state -->
<Style Selector="^:pointerover:not(.no-material,.NoFeedback) /template/ Border#PART_ButtonRootBorder">
    <Setter Property="BoxShadow" Value="{TemplateBinding Height, Converter={StaticResource ControlWidthToBoxShadowConverter}, ConverterParameter=1}"/>
    <Setter Property="Opacity" Value="1"/>
</Style>
```

Button shadow changed when overing the button with the mouse.

But with MultiBinding no overing anymore, i don't enter the converter with ConverterParameter=1 anymore :(


```
<Border Name="PART_ButtonRootBorder" Margin="{TemplateBinding Margin}" Padding="0"
                Width="{TemplateBinding Width}" Height="{TemplateBinding Height}" Cursor="{TemplateBinding Cursor}"
                VerticalAlignment="Center" HorizontalAlignment="Center"
                Background="{TemplateBinding Background}"
                BorderBrush="{TemplateBinding BorderBrush}"
                BorderThickness="{TemplateBinding BorderThickness}"
                CornerRadius="{TemplateBinding CornerRadius}">
                  <Border.BoxShadow>
                    <MultiBinding Converter="{StaticResource ControlWidthToBoxShadowConverter2}" ConverterParameter="0">
                        <Binding Path="Height" RelativeSource="{RelativeSource TemplatedParent}"/>
					    <DynamicResource ResourceKey="MaterialDesignShadowLightColor" />
					    <DynamicResource ResourceKey="MaterialDesignShadowDarkColor" />
                    </MultiBinding>
                  </Border.BoxShadow>

...

<!--Pointer enter state -->
<Style Selector="^:pointerover:not(.no-material,.NoFeedback) /template/ Border#PART_ButtonRootBorder">
    <Setter Property="BoxShadow">
    <Setter.Value>
        <MultiBinding Converter="{StaticResource ControlWidthToBoxShadowConverter2}" ConverterParameter="1">
            <Binding Path="Height" RelativeSource="{RelativeSource TemplatedParent}"/>
            <DynamicResource ResourceKey="MaterialDesignShadowLightColor" />
            <DynamicResource ResourceKey="MaterialDesignShadowDarkColor" />
        </MultiBinding>
    </Setter.Value>
    </Setter>
    <Setter Property="Opacity" Value="1"/>
</Style>
```

**Reply by timunie** [reply]
I think TemplatedParent is wrong here as the converter is not used inside ControlTemplate, right? Instead try $self of $parent[Border]

**Reply by flarive** [reply]
Hum, i don't think this is the problem...
TemplateBinding Height works well when used in a SingleBinding but not well when used into a MultiBinding.

And we are in a template context i think (/template/ Border#PART_ButtonRootBorder)

And i would receive some unset values in my converter if there were binding errors, that's not the case, i don't enter the converter when i mouse over my button.

I have tested your 2 suggestions but unfortunately they don't work.

**Reply by timunie** [reply]
can you prepare a very minimal sample of what you have so far? If it's a bug, we should then open an issue for it separately.

**Reply by flarive** [reply]
Hello Tim...
I have prepared a minimal sample here :
https://github.com/flarive/AvaloniaSampleDebug

The sample displays 2 buttons.
The first one works (simple binding) when you mouse over it, the second one doesn't work (multi binding)

Thanks a lot for your help :)

**Reply by timunie** [reply]
okay, I can reproduce your issue. @maxkatz6 should I open an issue for it? I believe a MulitConverter should work the same as a "nomral" converter

**Reply by flarive** [reply]
@maxkatz6 or @timunie can you open an issue for this problem ?
When does this MultiBinding issue should be resolved ?
Many thanks...

**Reply by timunie** [reply]
> When does this MultiBinding issue should be resolved? 

I cannot promise anything here. You're welcome to support us if you want to have it in fast and if you are able to fix.

**Reply by flarive** [reply]
Ok, create the issue and just tell me where multi binding is supposed to be managed in the Avalonia source code (class name and namespace) and i will try to have a look at it.
@timunie and @maxkatz6 I don't promise to fix it but i can try to help :)

**Comment by workgroupengineering** [comment, marked as answer]
I have similar issue. 

I corrected this way:

```xaml
  <ControlTheme x:Key="ButtonTheme2" TargetType="Button">
    <Setter Property="HorizontalContentAlignment" Value="Center" />
    <Setter Property="VerticalContentAlignment" Value="Center" />
    <Setter Property="Background" Value="Transparent" />
    <Setter Property="Cursor" Value="Hand" />
    <Setter Property="Padding" Value="16 8" />
    <Setter Property="Margin" Value="0" />
    <Setter Property="Foreground" Value="{DynamicResource MaterialDesignForeground}" />
    <Setter Property="FontSize" Value="15" />
    <Setter Property="FontWeight" Value="Normal"/>
    <Setter Property="FontStyle" Value="Normal"/>
    <Setter Property="ClipToBounds" Value="False" />
    <Setter Property="Width" Value="90" />
    <Setter Property="Height" Value="90" />
    <Setter Property="BorderBrush" Value="Transparent"/>
    <Setter Property="BorderThickness" Value="0"/>
    <Setter Property="CornerRadius" Value="0"/>
    <Setter Property="Template">
      <ControlTemplate>
        <Border Name="PART_ButtonRootBorder" Margin="{TemplateBinding Margin}" Padding="0"
            Width="{TemplateBinding Width}" Height="{TemplateBinding Height}" Cursor="{TemplateBinding Cursor}"
            VerticalAlignment="Center" HorizontalAlignment="Center"
            Background="{TemplateBinding Background}"
            BorderBrush="{TemplateBinding BorderBrush}"
            BorderThickness="{TemplateBinding BorderThickness}"
            CornerRadius="{TemplateBinding CornerRadius}">
          <Panel>
            <Border Name="PART_HoverIndicator" BorderThickness="{TemplateBinding BorderThickness}" BorderBrush="{TemplateBinding BorderBrush}"
                HorizontalAlignment="Stretch" VerticalAlignment="Stretch"
                Background="{TemplateBinding Background}" Padding="{TemplateBinding Padding}" Width="{TemplateBinding Width}" Height="{TemplateBinding Height}"
                CornerRadius="10"/>

            <ContentPresenter Name="PART_ContentPresenter"
                      Content="{TemplateBinding Content}"
                      ContentTemplate="{TemplateBinding ContentTemplate}"
                      Padding="{TemplateBinding Padding}"
                      TextBlock.Foreground="{TemplateBinding Foreground}"
                      TextBlock.FontFamily="{TemplateBinding FontFamily}"
                      TextBlock.FontSize="{TemplateBinding FontSize}"
                      TextBlock.FontWeight="{TemplateBinding FontWeight}"
                      TextBlock.FontStyle="{TemplateBinding FontStyle}"
                      HorizontalContentAlignment="{TemplateBinding HorizontalContentAlignment}"
                      VerticalContentAlignment="{TemplateBinding VerticalContentAlignment}"/>
          </Panel>
        </Border>
      </ControlTemplate>
    </Setter>


    <Style Selector="^:not(.no-material,.NoFeedback) /template/ Border#PART_ButtonRootBorder">
      <Setter Property="BoxShadow">
        <Setter.Value>
          <MultiBinding Converter="{StaticResource ControlWidthToBoxShadowConverter2}" ConverterParameter="0">
            <Binding Path="Height" RelativeSource="{RelativeSource TemplatedParent}"/>
            <DynamicResource ResourceKey="MaterialDesignShadowLightColor" />
            <DynamicResource ResourceKey="MaterialDesignShadowDarkColor" />
          </MultiBinding>
        </Setter.Value>
      </Setter>
      <Setter Property="Opacity" Value="1"/>
    </Style>

    <!--Pointer enter state -->
    <Style Selector="^:pointerover:not(.no-material,.NoFeedback) /template/ Border#PART_ButtonRootBorder">
      <Setter Property="BoxShadow">
        <Setter.Value>
          <MultiBinding Converter="{StaticResource ControlWidthToBoxShadowConverter2}" ConverterParameter="1">
            <Binding Path="Height" RelativeSource="{RelativeSource TemplatedParent}"/>
            <DynamicResource ResourceKey="MaterialDesignShadowLightColor" />
            <DynamicResource ResourceKey="MaterialDesignShadowDarkColor" />
          </MultiBinding>
        </Setter.Value>
      </Setter>
      <Setter Property="Opacity" Value="1"/>
    </Style>
  </ControlTheme>
```

I have move from default BoxShadow from ControlTheme.Template to Selector.

From what I understand the ControlTheme's ControlTemplate has a higher priority than the selector.

I don't know if that's the correct behavior.

What do you think @maxkatz6?

**Reply by flarive** [reply]
Yeah ! It works ! Thanks a lot for the hack :)
